cli/cmd/endpoints.go

```diff
@@ -0,0 +1,238 @@
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/linkerd/linkerd2/controller/api/public"
+	"github.com/linkerd/linkerd2/controller/gen/controller/discovery"
+	"github.com/linkerd/linkerd2/pkg/addr"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type endpointsOptions struct {
+	namespace    string
+	outputFormat string
+}
+
+var (
+	podHeader = "POD"
+)
+
+// validate performs all validation on the command-line options.
+// It returns the first error encountered, or `nil` if the options are valid.
+func (o *endpointsOptions) validate() error {
+	switch o.outputFormat {
+	case "table", "json", "":
+		return nil
+	}
+
+	return errors.New("--output currently only supports table and json")
+}
+
+func newEndpointsOptions() *endpointsOptions {
+	return &endpointsOptions{
+		namespace:    "",
+		outputFormat: "",
+	}
+}
+
+func newCmdEndpoints() *cobra.Command {
+	options := newEndpointsOptions()
+
+	cmd := &cobra.Command{
+		Use:   "endpoints [flags]",
+		Short: "Display service discovery endpoints state",
+		Long:  `Display service discovery endpoints state.`,
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := options.validate()
+			if err != nil {
+				return err
+			}
+
+			endpoints, err := requestEndpointsFromAPI(cliPublicAPIClient())
+			if err != nil {
+				return fmt.Errorf("Endpoints API error: %s", err)
+			}
+
+			output := renderEndpoints(endpoints, options)
+			_, err = fmt.Print(output)
+
+			return nil
+		},
+	}
+
+	cmd.PersistentFlags().StringVarP(&options.namespace, "namespace", "n", options.namespace, "Namespace of the specified endpoints (default: all namespaces)")
+	cmd.PersistentFlags().StringVarP(&options.outputFormat, "output", "o", options.outputFormat, "Output format; currently only \"table\" and \"json\" are supported (default \"table\")")
+
+	return cmd
+}
+
+func requestEndpointsFromAPI(client public.APIClient) (*discovery.EndpointsResponse, error) {
+	return client.Endpoints(context.Background(), &discovery.EndpointsParams{})
+}
+
+func renderEndpoints(endpoints *discovery.EndpointsResponse, options *endpointsOptions) string {
+	var buffer bytes.Buffer
+	w := tabwriter.NewWriter(&buffer, 0, 0, padding, ' ', 0)
+	writeEndpointsToBuffer(endpoints, w, options)
+	w.Flush()
+
+	return string(buffer.Bytes())
+}
+
+type rowEndpoint struct {
+	Namespace string `json:"namespace"`
+	IP        string `json:"ip"`
+	Port      uint32 `json:"port"`
+	Pod       string `json:"pod"`
+	Version   string `json:"version"`
+	Weight    uint32 `json:"weight"`
+	Identity  string `json:"identity"`
```

Comment:
Should this be called "service" or "authority"?  I think "identity" will soon come to mean TLS identity which will correspond to service account.